### PR TITLE
Tests for unit conversion

### PR DESCRIPTION
**Pull request by MichaelClerx**
See #122 

- [x] Input unit conversion: Setting initial value on existing state
- [x] Input unit conversion: Setting initial value on constant
- [x] Input unit conversion: Using define to change a constant
- [x] Input unit conversion: Using define to change an ODE

### CODE REVIEW

**Comment by MichaelClerx** [comment]
Hi @jonc125 , I think these are sensible tests. Could you double-check ?

**Comment by jonc125** [comment]
I'm surprised they pass - didn't realise we'd implemented all of these already!

File: test/protocols/test_unit_conversion_inputs_defines.txt [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    define diff(oxmeta:membrane_voltage; oxmeta:time) = 3 :: V_per_ms
+
+    # Change the value of n to 5M/ms
+    input test:parameter_n units M_per_ms
+    define test:parameter_n = 5 :: M_per_ms
+    output oxmeta:cytosolic_sodium_concentration units mM
+}
+
+tasks {
+    # Simulate for 10 seconds, storing only the first and last points
```
Comment (jonc125) [inline test/protocols/test_unit_conversion_inputs_defines.txt]: Actually 5 milliseconds...

File: test/protocols/test_unit_conversion_inputs_defines.txt [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    mV = milli volt
+    M = mole . litre^-1
+    mM = milli M
+    ms = milli second
+    V_per_ms = V . ms^-1
+    M_per_ms = M . ms^-1
+}
+
+model interface {
+    # Change the ODE for V to be in 3V/ms
```
Comment (jonc125) [inline test/protocols/test_unit_conversion_inputs_defines.txt]: If I'm being pedantic, cut 'in' :)

File: test/test_unit_conversion_inputs.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
+"""
+Tests for unit conversion of inputs and defines.
+For further unit conversion tests see test_graphstate.py
+"""
+import pytest
+
+import fc
+
+
+@pytest.mark.xfail(strict=True, reason='Unit conversion not implemented yet.')
```
Comment (jonc125) [inline test/test_unit_conversion_inputs.py]: Ah, that's why they pass :)

File: test/test_unit_conversion_inputs.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    proto = fc.Protocol(proto_file)
+    proto.set_output_folder('test_unit_conversion_inputs_initial_values')
+    proto.set_model('test/data/simple_ode.cellml')
+    proto.run()
+
+    # Check initial value of constant was set correctly
+    # The model has Na in mM with Na(0)=0mM, and dNa/dt=n, where n has initial value 1 mM/ms
+    # The protocol will change this to 3 M/ms, and then run for 5ms
+    Na = proto.output_env.look_up('cytosolic_sodium_concentration').array
+    assert len(Na) == 2
```
Comment (jonc125) [inline test/test_unit_conversion_inputs.py]: I'd be inclined to write this as `assert Na = [0, 15000]` as then you get more informative output from pytest if it fails.

File: test/protocols/test_unit_conversion_inputs_defines.txt [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+tasks {
+    # Simulate for 10 seconds, storing only the first and last points
+    simulation sim = timecourse {
+        range time units ms uniform 0:5:5
+    }
+}
+
+post-processing {
+    # Check dimensions of output
```
Comment (jonc125) [inline test/protocols/test_unit_conversion_inputs_defines.txt]: Maybe add a comment to say that the values will be checked in the test?

File: test/protocols/test_unit_conversion_inputs_initial_values.txt [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    input oxmeta:membrane_voltage units V = 10
+    output oxmeta:membrane_voltage units V
+
+    # Change the initial value of n to 3M/ms
+    input test:parameter_n units M_per_ms = 3
+    output oxmeta:cytosolic_sodium_concentration mM
+}
+
+tasks {
+    # Simulate for 10 seconds, storing only the first and last points
```
Comment (jonc125) [inline test/protocols/test_unit_conversion_inputs_initial_values.txt]: Again, 5ms

File: test/test_unit_conversion_inputs.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+    proto_file = 'test/protocols/test_unit_conversion_inputs_initial_values.txt'
+    proto = fc.Protocol(proto_file)
+    proto.set_output_folder('test_unit_conversion_inputs_initial_values')
+    proto.set_model('test/data/simple_ode.cellml')
+    proto.run()
+
+    # Check initial value of constant was set correctly
+    # The model has Na in mM with Na(0)=0mM, and dNa/dt=n, where n has initial value 1 mM/ms
+    # The protocol will change this to 3 M/ms, and then run for 5ms
```
Comment (jonc125) [inline test/test_unit_conversion_inputs.py]: 'this' -> 'n'

**Comment by MichaelClerx** [comment]
The don't pass! They're marked to xfail, as this was a PR for just the tests :D

File: test/test_unit_conversion_inputs.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    Na = proto.output_env.look_up('cytosolic_sodium_concentration').array
+    assert len(Na) == 2
+    assert Na[0] == 0
+    assert Na[1] == 15000
+
+    # Check initial value of state was set correctly
+    # The model has v in mV, with dv/dt=1 mV/ms, and v(0)=2mV
+    # The protocol will change this to v(0)=10 V, and then run for 5ms
```
Comment (jonc125) [inline test/test_unit_conversion_inputs.py]: Maybe also add that the output V is then in volts?

File: test/test_unit_conversion_inputs.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    v = proto.output_env.look_up('membrane_voltage').array
+    assert len(v) == 2
+    assert v[0] == 10
+    assert v[1] == 10.005
+
+
+@pytest.mark.xfail(strict=True, reason='Unit conversion not implemented yet.')
+def test_unit_conversion_inputs_defines():
+    # Test unit conversion for initial values of states and constants.
```
Comment (jonc125) [inline test/test_unit_conversion_inputs.py]: Copy/paste error?

File: test/test_unit_conversion_inputs.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+    proto_file = 'test/protocols/test_unit_conversion_inputs_defines.txt'
+    proto = fc.Protocol(proto_file)
+    proto.set_output_folder('test_unit_conversion_inputs_defines')
+    proto.set_model('test/data/simple_ode.cellml')
+    proto.run()
+
+    # Check initial value of constant was set correctly
+    # The model has Na in mM with Na(0)=0mM, and dNa/dt=n, where n has initial value 1 mM/ms
+    # The protocol will change this to 5 M/ms, and then run for 5ms
```
Comment (jonc125) [inline test/test_unit_conversion_inputs.py]: this -> n

File: test/test_unit_conversion_inputs.py [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+    Na = proto.output_env.look_up('cytosolic_sodium_concentration').array
+    assert len(Na) == 2
+    assert Na[0] == 0
+    assert Na[1] == 25000
+
+    # Check initial value of state was set correctly
+    # The model has v in mV, with dv/dt=1 mV/ms, and v(0)=2mV
+    # The protocol will change this to dv/dt=3 V/ms, and then run for 5ms
```
Comment (jonc125) [inline test/test_unit_conversion_inputs.py]: Again mention that the protocol reports V in volts.

File: test/protocols/test_unit_conversion_inputs_defines.txt (hunk 5 above)
Comment (MichaelClerx) [reply test/protocols/test_unit_conversion_inputs_defines.txt]: Good idea!

**Pull request merged**